### DISCUSSION TITLE
Open Links in a new Tab on iOS

### DISCUSSION THREAD

**Discussion by klubdergutenseiten-user** [SafeExamBrowser/seb-mac#215] [Q&A] [answered]
Hey All
I have a regular exam scenario where you are allowed to open links to other sites. The links are all in the whitelist and this works all very well, **but** I remember that on iOS it was possible to have multiple "tabs" in the left sidebar at the same time. When I open a link `<a href="example.com" target="_blank">Link</a>` it replaces the current exam browser window. It works on macOS, a new window is openened on top of the current exam window.

Expected / wished behaviour: a link openend in a new window (target _blank) should open a new tab on iOS. How is this possible?

Thanks for any hints!

Regards 
~flo 

Infos 
iOS 15.5
SEB (iOS) 3.0.1

Using this config-file: https://gist.github.com/klubdergutenseiten-user/a732e0cbc1cfb9485e7f8b3c0e9e8a51

**Comment by danschlet** [comment]
Generally opening new tabs works perfectly. I also don't see any non-default settings in your config file which would prevent that ("open in same tab", "block when on different server", URL filters).

Maybe your system uses some strange AJAX Javascript to open new tabs which might not be compatible with the classic WebView (which is very basic on iOS). If you would use a regular href-link with target="_blank", then it has to work, just try it with any simple HTML page or regular websites (I actually never came across any website where it wouldn't work).

You should also make your system compatible with the modern WebKit browser engine supported by SEB for iOS/macOS 3.0 and newer by using the [SEB JavaScript API to query the Config Key or Browser Exam Key](https://safeexambrowser.org/developer/seb-config-key.html). Especially on iOS the modern WebView is much better (also more compatible with complex web applications) than the classic UIWebView.

**Comment by klubdergutenseiten-user** [comment]
Ok, i found out, that when i set the option "Activate URL Filtering", new links are opened in the same window, even though they are regular links `<a href="example.com" target="_blank">Test</a>`
when i disable this setting, links are opened correctly in a new "tab". Is this a bug or can this be configured in the settings somehow?

My case is: I have some allowed urls, they should be opened in a new tab. This works on macOS by the way, since there a new window is opened on top of the exam window.

Thanks for your help!
~flo
![IMG_5024](https://user-images.githubusercontent.com/20813310/171659455-c79e0fc5-ceeb-485e-8d52-8e0674b636bc.jpg)

**Reply by danschlet** [reply]
I didn't try it with SEB 3.0.1 anymore, but SEB 3.1 actually crashes when a link is clicked which is supposed to be blocked when the URL filter is active (when using the modern WebView, with the classic one it seems to work). I have to debug that. Thank you very much for the bug report, now I can still fix this before releasing 3.1 (I was already submitting it to the App Store...).

**Comment by danschlet** [comment, marked as answer]
This should work correctly now in the just released SEB 3.1.

**Reply by klubdergutenseiten-user** [reply]
nice! i tested it and it works, thanks a lot, this is a small but really nice enhancement!

**Comment by klubdergutenseiten-user** [comment]
Hi Dan
I encountered the same issue again. if i restart my iPad, the link opens in a new window, if i redo the exam, the link opens in the same window. 

if i have a url without http**s** like http://google.com the site opens in the same window, on my older test iPad (air) it doesn't open at all.

Can i do something to help you investigate this issue? send you a debug log or does it help if i make a movie of the situation?

regards, flo

**Reply by danschlet** [reply]
Is it again related with URL filters?

I'm sorry, but I can only have a look at this issue if I have settings and access to a test site/account where I can reproduce it. I would need the exact steps to reproduce the problem. Sometimes also taking a video to demonstrate it helps. You can send confidential access data etc. to info at safeexambrowser.org (but we can't provide support over email).

**Comment by klubdergutenseiten-user** [comment]
sure, i'll send you some infos to your email. thx!

**Reply by danschlet** [reply]
So the reason why the links sometimes don't get opened in a new tab is a bit complicated:

The classic WebView (UIWebView) in iOS is really very, very primitive. For example its API doesn't support natively to open links in new tabs (or better said the app doesn't get an indication that the link should be opened in a new tab). The workaround (for all apps which are/were using UIWebView) is, that the app (SEB in this case) has to run a javascript which prefixes all links on a web page with target = _blank with a custom URL protocol scheme, so the link is modified for example to `newtab:https%3A//www.google.com/` (there is some escaping necessary hence the "%3A"). When the user taps the link, SEB detects that it's a "newtab:" link and knows it should be opened in a new tab.

The js script which modifies each page has to run after the page source has been completely loaded, it is executed in a delegate (event handler) method called "webViewDidFinishLoad". For some reason in some cases the page source is already loaded completely when webViewDidFinishLoad is called, sometimes it isn't. In the latter case the links are not getting prefixed with "newtab:" and therefore SEB doesn't know that they should be opened in a new tab. 

I don't know why the page source sometimes isn't loaded completely when webViewDidFinishLoad is called. As you said, when SEB is started the first time, then it works, if you do one test run, exit the exam (session) in SEB and start a second one, without manually terminating SEB, then it doesn't work. In the second case when you do a page reload, then it works correctly again (the links are prefixed with newtab and open in new tabs).

I couldn't reproduce this with a simple HTML page like https://demo.safeexambrowser.org/exams/ (the links at the bottom open in new tabs) and as mentioned I never observed it before (and didn't get such feedback, besides in one other case), so I assume the reason it happens with your exam system is that you're using some special way how to load the page/populate the DOM. Maybe it is a caching issue in SEB (after the first run the page is maybe already cached in the browser engine for SEB). It could also be the web server you're using, but again if it would happen in many cases I surely would have gotten such feedback before. In that one case I got a report in the past they added the links with AJAX after the page was loaded.

You can either try change the way you're loading/populating the page. But the better way would be to make sure that the modern WebView can be used by SEB, because that doesn't have such restrictions, is much faster, supports modern Web APIs and is more stable (as the content is rendered in separate processes and doesn't crash SEB if it the renderer crashes). See the link I posted above. In short because you have the setting sendBrowserExamKey = true, SEB uses the classic WebView. If you're using the ConfigKey or BEK, you need to query them using the SEB JavaScript API, see the mentioned developer documentation.
